Add rpc.functions.async_execution decorator for rpc_sync/rpc_async

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #39267 Add rpc.async_function decorator for TorchScript functions
* **#39216 Add rpc.functions.async_execution decorator for rpc_sync/rpc_async**

The `rpc.functions.async_execution` decorator specifies that the
wrapped function is guaranteed to return a `torch.futures.Future`.
The decorator adds a `_wrapped_async_rpc_function` attribute to
the wrapper function. The caller retrieves this information and
then sets `isAsyncFunction` argument accordingly which is later
added to PythonCall RPC message as a field. On the callee side,
if the PythonCall carries an asynchronous function, it will cast
the function's return value to a jit::PythonFutureWrapper object,
and then install response creation and communication as a callback
on the that jit::PythonFutureWrapper.

For applications, this feature is useful when a function needs to
wait for IO or additional singaling. In those cases, marking the
user function as `rpc.functions.async_execution` will prevent it
from blocking one thread on callee for too long.

Differential Revision: [D21779962](https://our.internmc.facebook.com/intern/diff/D21779962)

mrshenli committed:
Add rpc.async_function decorator for rpc_async

[ghstack-poisoned]

mrshenli commented on torch/distributed/rpc/api.py:
```diff
@@ -294,6 +294,18 @@ def _init_rpc_backend(
+    @functools.wraps(fn)
+    def wrapper(*args, **kwargs):
+        return fn(*args, **kwargs)
+    wrapper._wrapped_async_rpc_function = fn
```
I am not a fan of adding one attribute to distinguish whether this is an async function. Is there a better way to achieve this? We basically just need to know if a given function is wrapped by `@rpc.async_function`.

rohan-varma replied:
I'm not sure, though do we need to set this to `fn`? It looks like we just use it as a boolean check, could we just use `True` and call it `_is_async_fn`?

mrshenli replied:
It's used in the #39267 to retrieve the TorchScript function.

mrshenli committed:
Update on "[WIP] Add rpc.async_function decorator for rpc_async"




[ghstack-poisoned]

mrshenli committed:
Update on "[WIP] Add rpc.async_function decorator for rpc_async"


Differential Revision: [D21779962](https://our.internmc.facebook.com/intern/diff/D21779962)

[ghstack-poisoned]

mrshenli committed:
Update on "[WIP] Add rpc.async_function decorator for rpc_async"


Differential Revision: [D21779962](https://our.internmc.facebook.com/intern/diff/D21779962)

[ghstack-poisoned]

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -339,6 +339,70 @@ def fail_on_fut(fut):
     pass
 
 
+@rpc.async_function
```
I wonder if we should name this `@rpc.functions.returns_future` instead. In this way, we can mark different types of functions later, e.g., `@rpc.functions.batch`, `@rpc.function.fifo`, etc.

pritamdamania87 replied:
I would rather prefer `@rpc.function.async` if we're planning to have a `@rpc.function.*` group.

mrshenli replied:
Thanks, will update

mrshenli replied:
@pritamdamania `async` collides with Python keyword. I will name it `async_execution` for now. Feel free to comment if the name does not look good.

mrshenli committed:
Update on "Add rpc.async_function decorator for rpc_async"


Differential Revision: [D21779962](https://our.internmc.facebook.com/intern/diff/D21779962)

[ghstack-poisoned]

mrshenli committed:
Update on "Add rpc.async_function decorator for rpc_async"


Differential Revision: [D21779962](https://our.internmc.facebook.com/intern/diff/D21779962)

[ghstack-poisoned]

mrshenli committed:
Update on "Add rpc.async_function decorator for rpc_async"


Differential Revision: [D21779962](https://our.internmc.facebook.com/intern/diff/D21779962)

[ghstack-poisoned]

rohan-varma commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -168,15 +169,70 @@ void RequestCallbackImpl::processRpc(
+            return;
+          }
+        }
+        pyFuture->fut->addCallback([msgId = messageId,
```
So, with this change, functions decorated with `@rpc.rpc_async` will no longer block a thread on the server, correct? Do we need to update documentation anywhere to recommend users to use this in case their UDFs involve RPC or the like?

mrshenli replied:
Good point! Will add the doc change to `rpc.rst` when the PR for `rpc.remote` is also ready. If we add the change to API docs now, it will make `rpc_async`, `rpc_sync`, and `remote` inconsistent.

mrshenli replied:
Created #39399 to track

rohan-varma commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -168,15 +169,70 @@ void RequestCallbackImpl::processRpc(
+          responseFuture->markCompleted(std::move(m));
+        });
+      } else {
+        std::shared_ptr<SerializedPyObj> serializedPyObj;
```
What happens if a UDF returns a future that's set later, but is not decorated properly? (I guess this q is even valid today, when we don't have this decorator). Do we need error handling for this?

mrshenli replied:
In that case, the RPC would try to pickle the Future, which would hit the pickle error. Let me add a test for that.

pritamdamania87 commented:
Can we add a detailed summary for this PR?

pritamdamania87 commented on torch/distributed/rpc/api.py:
```diff
@@ -294,6 +294,18 @@ def _init_rpc_backend(
+def async_function(fn):
+    r"""
+    Wrapper of a function indicating that the return value of the function is
+    guranteed to be a ``torch.futures.Future``.
```
nit: guaranteed

Can we add some detailed documentation here in terms of the purpose of this decorator and how it is supposed to be used? Also, we should probably mention this is a decorator instead of saying "Wrapper of a function" ?

pritamdamania87 commented on torch/csrc/distributed/rpc/python_call.cpp:
```diff
@@ -6,23 +6,34 @@ namespace torch {
+  const char& c = message.payload()[0];
+  TORCH_INTERNAL_ASSERT(c == 0 || c == 1);
+  bool isAsyncFunction = (c == 1);
+  std::string payload(message.payload().begin() + 1, message.payload().end());
```
nit: add a DCHECK that the payload is at least of size 1?

pritamdamania87 commented on torch/csrc/distributed/rpc/python_functions.cpp:
```diff
@@ -173,11 +173,13 @@ c10::intrusive_ptr<JitFuture> pyRpcPythonUdf(
     std::vector<torch::Tensor>& tensors,
-    const float rpcTimeoutSeconds) {
+    const float rpcTimeoutSeconds,
+    const bool asyncFunction) {
```
nit: probably use `isAsyncFunction` or `asyncFunction` consistently?

pritamdamania87 commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -168,15 +169,70 @@ void RequestCallbackImpl::processRpc(
-        serializedPyObj =
-            std::make_shared<SerializedPyObj>(pythonRpcHandler.serialize(
-                pythonRpcHandler.runPythonUdf(std::move(upc).movePythonUdf())));
+      if (upc.isAsyncFunction()) {
```
Do we need to send the `isAsync` bit over the wire? Can't we inspect the pyobject here to see if it has a `_wrapped_async_rpc_function` attribute?

pritamdamania87 commented on torch/csrc/distributed/rpc/python_rpc_handler.cpp:
```diff
@@ -158,6 +160,15 @@ void PythonRpcHandler::handleExceptionGILHeld(const py::object& obj) {
   pyHandleException_(obj);
 }
 
+bool PythonRpcHandler::isException(const py::object& obj) {
```
nit: Should this be called `isRemoteException`?

pritamdamania87 commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -168,15 +169,70 @@ void RequestCallbackImpl::processRpc(
+
+          try {
+            pyFuture = result.cast<std::shared_ptr<jit::PythonFutureWrapper>>();
+          } catch (...) {
```
What sort of exception do we typically get here when the cast fails? Shouldn't catching `py::cast_error` be sufficient here?

mrshenli replied:
Good point, that should be should be sufficient. And other types of errors should propagate to upper layer try-catch in `processMessage`.

pritamdamania87 reviewed (changes requested):
Awesome work, really excited to see the async framework coming together! Requesting changes since I think it should be possible to avoid sending the async bit over the wire?

pritamdamania87 commented:
Actually after looking at https://github.com/pytorch/pytorch/pull/39267, I think sending the async bit over is better. Since we if look this up on the server side, that would incur Python GIL in the torchscript case.

mrshenli committed:
Update on "Add rpc.async_function decorator for rpc_async"


Differential Revision: [D21779962](https://our.internmc.facebook.com/intern/diff/D21779962)

[ghstack-poisoned]

mrshenli committed:
Update on "Add rpc.async_function decorator for rpc_async"


Differential Revision: [D21779962](https://our.internmc.facebook.com/intern/diff/D21779962)

[ghstack-poisoned]

mrshenli committed:
Update on "Add rpc.async_function decorator for rpc_async"


Differential Revision: [D21779962](https://our.internmc.facebook.com/intern/diff/D21779962)

[ghstack-poisoned]

mrshenli committed:
Update on "Add rpc.async_function decorator for rpc_async"


Differential Revision: [D21779962](https://our.internmc.facebook.com/intern/diff/D21779962)

[ghstack-poisoned]

mrshenli committed:
Update on "Add rpc.async_function decorator for rpc_async"


Differential Revision: [D21779962](https://our.internmc.facebook.com/intern/diff/D21779962)

[ghstack-poisoned]

mrshenli committed:
Update on "Add rpc.async_function decorator for rpc_async"


Differential Revision: [D21779962](https://our.internmc.facebook.com/intern/diff/D21779962)

[ghstack-poisoned]

mrshenli committed:
Update on "Add rpc.functions.async_execution decorator for rpc_sync/rpc_async"


The `rpc.functions.async_execution` decorator specifies that the
wrapped function is guaranteed to return a `torch.futures.Future`.
The decorator adds a `_wrapped_async_rpc_function` attribute to
the wrapper function. The caller retrieves this information and
then sets `isAsyncFunction` argument accordingly which is later
added to PythonCall RPC message as a field. On the callee side,
if the PythonCall carries an asynchronous function, it will cast
the function's return value to a jit::PythonFutureWrapper object,
and then install response creation and communication as a callback
on the that jit::PythonFutureWrapper.

For applications, this feature is useful when a function needs to
wait for IO or additional singaling. In those cases, marking the
user function as `rpc.functions.async_execution` will prevent it
from blocking one thread on callee for too long.

Differential Revision: [D21779962](https://our.internmc.facebook.com/intern/diff/D21779962)

[ghstack-poisoned]

mrshenli committed:
Update on "Add rpc.functions.async_execution decorator for rpc_sync/rpc_async"


The `rpc.functions.async_execution` decorator specifies that the
wrapped function is guaranteed to return a `torch.futures.Future`.
The decorator adds a `_wrapped_async_rpc_function` attribute to
the wrapper function. The caller retrieves this information and
then sets `isAsyncFunction` argument accordingly which is later
added to PythonCall RPC message as a field. On the callee side,
if the PythonCall carries an asynchronous function, it will cast
the function's return value to a jit::PythonFutureWrapper object,
and then install response creation and communication as a callback
on the that jit::PythonFutureWrapper.

For applications, this feature is useful when a function needs to
wait for IO or additional singaling. In those cases, marking the
user function as `rpc.functions.async_execution` will prevent it
from blocking one thread on callee for too long.

Differential Revision: [D21779962](https://our.internmc.facebook.com/intern/diff/D21779962)

[ghstack-poisoned]

mrshenli committed:
Update on "Add rpc.functions.async_execution decorator for rpc_sync/rpc_async"


The `rpc.functions.async_execution` decorator specifies that the
wrapped function is guaranteed to return a `torch.futures.Future`.
The decorator adds a `_wrapped_async_rpc_function` attribute to
the wrapper function. The caller retrieves this information and
then sets `isAsyncFunction` argument accordingly which is later
added to PythonCall RPC message as a field. On the callee side,
if the PythonCall carries an asynchronous function, it will cast
the function's return value to a jit::PythonFutureWrapper object,
and then install response creation and communication as a callback
on the that jit::PythonFutureWrapper.

For applications, this feature is useful when a function needs to
wait for IO or additional singaling. In those cases, marking the
user function as `rpc.functions.async_execution` will prevent it
from blocking one thread on callee for too long.

Differential Revision: [D21779962](https://our.internmc.facebook.com/intern/diff/D21779962)

[ghstack-poisoned]

pritamdamania87 commented on torch/distributed/rpc/functions.py:
```diff
@@ -0,0 +1,46 @@
+    is guaranteed to be a ``torch.futures.Future`` object and this function can
+    run asynchronously on the RPC callee. This is useful when this function's
+    execution needs to pause and resume due to, e.g., containing
+    :meth:`~torch.distributed.rpc.rpc_async` or waiting for other signals.
```
Should we add that the RPC on the callee side responds to the caller only when the returned future is completed? This would make it clear that we're not actually returning a Future over the wire.

Also, should we mention the order of decorators here? Basically this one needs to be the outer most decorator.

mrshenli committed:
Update on "Add rpc.functions.async_execution decorator for rpc_sync/rpc_async"


The `rpc.functions.async_execution` decorator specifies that the
wrapped function is guaranteed to return a `torch.futures.Future`.
The decorator adds a `_wrapped_async_rpc_function` attribute to
the wrapper function. The caller retrieves this information and
then sets `isAsyncFunction` argument accordingly which is later
added to PythonCall RPC message as a field. On the callee side,
if the PythonCall carries an asynchronous function, it will cast
the function's return value to a jit::PythonFutureWrapper object,
and then install response creation and communication as a callback
on the that jit::PythonFutureWrapper.

For applications, this feature is useful when a function needs to
wait for IO or additional singaling. In those cases, marking the
user function as `rpc.functions.async_execution` will prevent it
from blocking one thread on callee for too long.

Differential Revision: [D21779962](https://our.internmc.facebook.com/intern/diff/D21779962)

[ghstack-poisoned]

mrshenli commented:
Docs looking OK. Landing now

<img width="855" alt="Screen Shot 2020-06-02 at 10 02 39 PM" src="https://user-images.githubusercontent.com/16999635/83587706-b5be3e80-a51d-11ea-8975-93183b19f286.png">

mrshenli committed:
Update on "Add rpc.functions.async_execution decorator for rpc_sync/rpc_async"


The `rpc.functions.async_execution` decorator specifies that the
wrapped function is guaranteed to return a `torch.futures.Future`.
The decorator adds a `_wrapped_async_rpc_function` attribute to
the wrapper function. The caller retrieves this information and
then sets `isAsyncFunction` argument accordingly which is later
added to PythonCall RPC message as a field. On the callee side,
if the PythonCall carries an asynchronous function, it will cast
the function's return value to a jit::PythonFutureWrapper object,
and then install response creation and communication as a callback
on the that jit::PythonFutureWrapper.

For applications, this feature is useful when a function needs to
wait for IO or additional singaling. In those cases, marking the
user function as `rpc.functions.async_execution` will prevent it
from blocking one thread on callee for too long.

Differential Revision: [D21779962](https://our.internmc.facebook.com/intern/diff/D21779962)

[ghstack-poisoned]
